book/invalidation.md

```diff
@@ -0,0 +1,379 @@
+
+``` {.python}
+class DocumentLayout:
+    def layout(self, zoom):
+        if not self.children:
+            child = BlockLayout(self.node, self, None)
+        else:
+            child = self.children[0]
+        self.children = [child]
+        # ...
+```
+
+But we have similar issues in `BlockLayout`, which creates all the
+other types of layout objects and `append`s them to its `children`
+array. Here, the easy fix is to reset the `children` array at the top
+of `layout`:
+
+``` {.python}
+class BlockLayout:
+    def layout(self, zoom):
+        self.children = []
+        # ...
+```
+
+Now the `BlockLayout`'s `layout` function is idempotent again, because
+each call will recreate the `children` array the same way each time.
+
+I examined all of the other `layout` methods to check whether they are
+idempotent. I found a few worrying causes:
+
+- In `new_line`, `BlockLayout` will append to its `children` array;
+- In `text` and `input`, `BlockLayout` will append to the `children`
+  array of some `LineLayout` child
+- In `text` and `input`, `BlockLayout` will call `get_font`, as will
+  the `TextLayout` and `InputLayout` methods
+- Basically every layout method calls `display_px`
+
+Luckily, none of these break idempotency. The `new_line` method is
+only called through `layout`, which resets the `children` array.
+Similarly, `text` and `input` are also only called through `layout`,
+which means all of the `LineLayout` children are reset. And `get_font`
+acts as a cache, so multiple calls return the same font object, while
+`display_px` just does some math so always returns the same result
+given the same inputs. This means the browser should now work
+correctly again.
+
+::: {.further}
+
+:::
+
+Dependencies
+============
+
+So far, we've only looked at one place where layout objects are
+created. Let's look at another: `BlockLayout`s created by
+`BlockLayout`s in their `layout` method. Here's the relevant code;
+it only runs in block layout mode:
+
+``` {.python}
+class BlockLayout:
+    def layout(self, zoom):
+        self.children = []
+        # ...
+        if layout_mode(self.node) == "block":
+            previous = None
+            for child in self.node.children:
+                next = BlockLayout(child, self, previous)
+                self.children.append(next)
+                previous = next
+        # ...
+```
+
+Here, the arguments to `BlockLayout` are a little more complicated
+than the ones passed in `DocumentLayout`. The second argument, `self`,
+is never assigned to, but `child` and `previous` both contain elements
+read from `node.children`, and that `children` array can change as a
+result of `innerHTML`. Moreover, in order to even run this code, the
+node's `layout_mode` has to be `"block"`, and `layout_mode` itself
+also reads the node's `children`.[^and-tags]
+
+[^and-tags]: It also looks at the node's `tag` and the node's
+    childrens' `tag`s, but luckily in our browser a node's `tag` can't
```

Comment:
random note: tags can't change in a real browser without destroying the DOM node.

Reply:
no such DOM API

Reply:
literally this (implicitly in the context of idempotence, without saying that word) came up in a design review for a new DOM API a month ago.